### ISSUE TITLE
Cumsum Creating Discontinuous Lines

### ISSUE THREAD

**Issue by whynotbrooklyn** [apache/superset#21093]
#### How to reproduce the bug

1. Create a cumsum with multiple series where some series do not have all of the dates
2. Plot Smooth Line

### Expected results

I would expect that when there are gaps in the date data, that the superset cumsum function creates a continuous line filling with the last value rather than producing NAs.

### Actual results

Lines break due to NAs in table when one series has a data point and the others don't.

#### Screenshots

![image](https://user-images.githubusercontent.com/104156928/184678320-9f00f687-2a1c-4f83-8737-637570c0c087.png)

### Environment

- Chrome latest version
- superset version: 1.5.0

**Comment by whynotbrooklyn** [comment]
I think it's related to this bug that was fixed for single series charts: https://github.com/apache/superset/issues/1661

**Comment by ryandawsonuk** [comment]
I'm also seeing the same problem with different styles of presenting the chart - seeing with both line and bar

**Comment by whynotbrooklyn** [comment]
Hi [rusackas](https://github.com/rusackas) is there any way to get some traction on this? My company is trying to utilize Superset for a commercial offering and this is a pretty glaring issue.

**Comment by sfirke** [comment]
Hi there - it's not a function of Superset I'd ever used before nor do I know its codebase.  So I can't speak to this, sorry,  But it seems like a well-done bug report.

**Comment by whynotbrooklyn** [comment]
> Hi there - it's not a function of Superset I'd ever used before nor do I know its codebase. So I can't speak to this, sorry, But it seems like a well-done bug report.

Well thanks for taking a look at the ticket anyway! Sorry to randomly tag; we're getting pretty desperate to get this fixed. Our client just asked for the cumulative charts in weekly rather than monthly increments - which only makes this bug look worse. 

Given that this has already been fixed for single series charts, I'm hoping it won't be an intensive to apply the same fix for the multi-series version. I'd be happy to even take a stab at it myself if anyone could point me to the right file location. 

Adding in @lilykuang @hughhhh and @michael-s-molina for expanded visibility.

**Comment by rusackas** [comment]
You can *probably* resolve this (though I don't have a dataset to repro handy) by using the Zero Imputation option in the Fill Method control of the Advanced Analytics section of chart controls. This will take any null datapoints and convert them to 0 values, which I would _assume_ means you'll get a contiguous line thanks to the now-contiguous dataset.

<img width="316" alt="image" src="https://user-images.githubusercontent.com/812905/215866023-b7afbe0d-5b3c-43cc-aeba-bbb5c1e5f67c.png">

**Comment by whynotbrooklyn** [comment]
Hi @rusackas , thanks for getting back to me!

Unfortunately, I have tried every one of the "FILL METHOD" options and none of them resolve the NAs in the subsequent output. (You can see the NAs in the table below the chart in my example pic.) It makes sense that this wouldn't have much effect given that there is no resampling done in my data. However, adding any resampling rules just produced errors so I'm not sure that is the answer either. How was this solved in the single series example that I referenced? 

I have mocked up some demo data for you so that this can be better triaged: [Discontinuous Cumsum Demo Data](https://github.com/apache/superset/files/10550914/2023-01-31.2_41pm.csv)

The options that I have selected are simply:
X-Axis: event_date
Metric: COUNT(*)
Dimensions: grouping_variable
Rolling Function: cumsum

**Comment by rusackas** [comment]
Thank you for the example dataset! Sadly, I was indeed able to validate the issue you're experiencing. I'll tag a few people who might take interest in this if/when time allows, but I'd love to see _anyone_ give this issue some help.

**Comment by sfirke** [comment]
A week later and now I'm facing this bug too.  If this is the same issue it's not a matter of cumsum specifically.  In my case I'm seeing zeroes correctly imputed for intervals when no dimension group has a record, but if dimension group _does_ have a record, the others don't get the zeroes.  Look at Feb 7th in the table and the chart of this screenshot:
![image](https://user-images.githubusercontent.com/7569808/218163835-c69cedc7-0d70-463e-b481-f3dac4f3d5f7.png)

**Comment by whynotbrooklyn** [comment]
@yousoph @zhaoyongjie @villebro Any update on when this can get fixed? 

We're supposed to be going live with this soon and this graph is pretty bad... especially when you notice that without markers, you can't even tell that there's data for May in the control group...
![no_markers](https://user-images.githubusercontent.com/104156928/225406211-9d3f3e8d-3d2e-4108-96af-e63bb25610e2.PNG)

![with_markers](https://user-images.githubusercontent.com/104156928/225406165-7d8ac23a-48f5-400f-badc-77902b639f37.PNG)

**Comment by whynotbrooklyn** [comment]
This one is hardly even a line graph anymore...
![image](https://user-images.githubusercontent.com/104156928/225407116-0f9e4f69-1e30-463b-8c78-30be5e90633f.png)

**Comment by sandeepson** [comment]
I am getting the same issue for the line chart. Following this thread for a resolution. Please assist!!

**Comment by HeathLee** [comment]
getting  same issue

**Comment by WulfH** [comment]
The fix is probably very simple. In superset\viz.py in this line https://github.com/apache/superset/blob/46a0a6e08af1b66bb617d5ebe1324b77251aee15/superset/viz.py#L232
the nan values should be replaced with 0. 
`df = df.fillna(0).cumsum()`

I can't easily test it, but the behavior is well documented: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cumsum.html

**Comment by lchavas** [comment]
@WulfH thank you, my cumsum dimensions (sales subdivided into stacked bars per family) were always truncated after the last value in the DataFrame, which resulted in stacked subdivisions not matching the total.
fillna(0) fixed my sums and the subtotals now match the main line, it's a beautiful sight to behold. This should be pulled

**Comment by whynotbrooklyn** [comment]
@rusackas is there anything you can do to get this change pulled into main?

**Comment by zhaoyongjie** [comment]
This seems to be a new feature,

It's not the same as the 'resample' operation. '[Resample](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.resample.html)' refers to filling in rows with missing time indices, whereas this feature requires modifying the original Dataframe which means replace some values with specific values, e.g. null -> 0.

**Comment by daro-123** [comment]
I have this issue as well where there is no continuous line on some dates when using cumsum but resample does the trick and help to fill in the gap. 
**Before** 
![image](https://github.com/apache/superset/assets/13133725/cb1e2023-8ea0-44df-8c67-e4805e2b679a)
**After**
![image](https://github.com/apache/superset/assets/13133725/c94de1ff-d013-4e49-8d67-f8acaf7fc396)
![image](https://github.com/apache/superset/assets/13133725/9011931c-5f35-401a-8432-19e340d7afb5)
My chart time grain is week; I'm on 2.1.1.

**Comment by diopol** [comment]
> The fix is probably very simple. In superset\viz.py in this line
> 
> https://github.com/apache/superset/blob/46a0a6e08af1b66bb617d5ebe1324b77251aee15/superset/viz.py#L232
> 
> 
> the nan values should be replaced with 0.
> `df = df.fillna(0).cumsum()`
> I can't easily test it, but the behavior is well documented: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cumsum.html

this fix works

**Comment by diopol** [comment]
this change worked for me:

https://github.com/apache/superset/blob/46a0a6e08af1b66bb617d5ebe1324b77251aee15/superset/utils/pandas_postprocessing/cum.py#L56

changed to:
```python
df_cum = _append_columns(df, getattr(df_cum.fillna(0), operation)(), columns)
```

I'm using `3.0.0rc3`

**Comment by villebro** [comment]
FYI here's a fix for this issue: #26429

**Comment by JohnDietrich-Pepper** [comment]
Ah this fix really broke our charting.  We wanted to show discontinuous lines when the data is null because the data is null and NOT zero particularly when the time period hasn't occurred yet.  Now Superset is assuming all null data is zero which is not the case.
@rusackas @villebro
